File: java17/src/main/java/folio/codinginterview/domain/StockSymbol.java
package folio.codinginterview.domain;

import java.util.Optional;

/** 銘柄を表す。 */
public enum StockSymbol {
    Toyopa, Somy;

    public static Optional<StockSymbol> fromString(String s) {
        return switch (s) {
            case "Toyopa" -> Optional.of(Toyopa);
            case "Somy" -> Optional.of(Somy);
            default -> Optional.empty();
        };
    }
}

File: golang/internal/domain/stock_symbol.go
package domain

import "fmt"

type StockSymbol string

const (
	Toyopa StockSymbol = "Toyopa"
	Somy   StockSymbol = "Somy"
)

func StockSymbolFromString(s string) (StockSymbol, error) {
	switch s {
	case "Toyopa":
		return Toyopa, nil
	case "Somy":
		return Somy, nil
	default:
		return "", fmt.Errorf("unknown symbol: %s", s)
	}
}
